Improve API access control for linked domains

gherceg opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
https://dimagi-dev.atlassian.net/browse/SAAS-13357

~I've been hesitant to create this PR because I'm not thrilled with this, but do think we need to enhance access control around API endpoints for linked domains. The main change is to check if the requesting user has access to linked domains in addition to just checking if the requesting domain is part of an active domain link.~

**Update** The API endpoints impacted by this change are responsible for retrieving data models from the upstream domain being queried. Previously, a check enforced that the domain being queried was part of active link with the domain specified in `REMOTE_REQUESTER_HEADER`. This is to ensure that the requesting user/domain has proper access to this data. In practice, this request is made by HQ directly, and this kind of check isn't necessary in that context so I'm assuming this check is strictly a security measure.

I've replaced this check with the standard check for access to linked domains, which opens up the possibility that a user with linked domains permissions in both domains could retrieve data models from a domain that is **not** part of an active link with the requesting domain. I'm inclined to think this isn't a risk since the user has access to this information given their linked domain permission. Does this sound particularly concerning to anyone?

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
Will test locally before merging. Want to get some opinions first.

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->


### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

gherceg committed:
attempt to fetch domain link directly

gherceg committed:
refactor decorator name

gherceg committed:
ensure user has access to linked domains

gherceg committed:
keep decorator responsibilities separate

gherceg commented:
Alright I decided what I didn't like was that I had the `require_existing_domain_link` decorator handling multiple responsibilities, rather than just adding an additional decorator to the API methods. These methods are used within HQ (see linked_domains/remote_accessors.py) so I feel confident that users of this endpoint will already have access to linked domains, and therefore not be disrupted by this change.

gherceg committed:
fix failing test by mocking linked domain access

mjriley commented on corehq/apps/linked_domain/models.py:
```diff
@@ -83,11 +83,13 @@ def update_last_pull(self, model, user_id, date=None, model_detail=None):
             is_active_downstream_domain,
             is_active_upstream_domain,
         )
+        get_domain_link.clear(self.master_domain, self.linked_domain)
```
This is theoretically error-prone. Consider a domain link between domains A and B. For whatever reason, someone pulls up this domain link and changes the link to be between A and C. This method will clear the cache for A and C, but the cache will still believe there is a valid link between A and B.

gherceg replied:
This has been resolved in that I've removed the `get_domain_link` method entirely, but this brings up an interesting thought. As far as available workflows to web users, a domain link cannot be edited in that way. One would need to open a db shell and edit the object directly to achieve this scenario. If something like this comes up in the future, should we build into code the ability to handle edge cases like this?

mjriley replied:
I don't think existing workflows are enough of a safeguard to prevent misuse in the future. We want to design our code to be robust, not knowing how it may be used in the future. There is definitely a valid argument about over-engineering, but if we feel that something like this would be over-engineering, we should at least be leaving comments in the code mentioning that domain links should not be modified in this way, and will break if they are.

mjriley reviewed (commented):
Nice tests! I like the thought behind these changes. I do want to make sure that decorators are the best approach here, however.

gherceg commented on corehq/apps/linked_domain/tests/test_remote_auth.py:
```diff
@@ -13,46 +12,62 @@
+            'HTTP_AUTHORIZATION': f'apikey test:{self.api_key.key}',
+            REMOTE_REQUESTER_HEADER: self.downstream_domain_requester,
+        }
+        with mock.patch('corehq.apps.linked_domain.decorators.can_user_access_linked_domains', return_value=False):
```
Interesting I didn't think of specifying the module as the object when using `patch.object`. I like that. I think my personal preference is to avoid specifying patches in setup that only apply to a subset of tests in a class, so in this case I prefer using `patch.object` over a move to setup. I'll stick with that unless you feel strongly the other way.

gherceg commented on corehq/apps/linked_domain/tests/test_remote_auth.py:
```diff
@@ -13,46 +12,62 @@
+
+    @classmethod
+    def setUpClass(cls):
+        super(RemoteAuthTest, cls).setUpClass()
```
Didn't catch this when moving this method to the bottom of the class. Thanks!

gherceg committed:
address testing feedback

gherceg committed:
remove require_existing_domain_link decorator

gherceg committed:
remove unnecessary dbaccessor method

gherceg commented:
After offlining with Matt, it seems that the move to class based views would be controversial, and in reality the best solution here is likely to just remove this decorator/check for an active domain link altogether, which will simplify things. I was just trying to keep existing code as is, but upon further reflection the improved check for linked domains access is more than enough from a security perspective. A user must have the linked domains permission enabled in the upstream remote domain they attempt to pull data from.

gherceg commented on corehq/apps/linked_domain/dbaccessors.py:
```diff
@@ -13,6 +13,19 @@
 from corehq.util.quickcache import quickcache
 
 
+@quickcache(['upstream', 'downstream'], timeout=60 * 60)
```
This has been resolved as I've removed `get_domain_link` entirely, but this is good point that unless there is a clear performance gain, it isn't worth adding the complexity of caching. Noted for the future!

orangejenny replied:
To pile on, also noting that caching python objects can add friction to future changes (e.g., renaming the model and needing to have code that's backwards-compatible so objects cached prior to the code deploy can still can be fetched).

gherceg commented on corehq/apps/linked_domain/views.py:
```diff
@@ -120,62 +120,72 @@
 @login_or_api_key
-@require_linked_domain
+@require_existing_domain_link
+@require_access_to_linked_domains
```
Ended up removing the `require_existing_domain_link` decorator entirely. I've justified this below, but don't see this as a necessary check now that we have user permission checks as part of `require_access_to_linked_domains`.

gherceg committed:
remove references to REMOTE_REQUESTER_HEADER and fix tests

gherceg commented:
Updated the description, but posting here as well:
> The API endpoints impacted by this change are responsible for retrieving data models from the upstream domain being queried. Previously, a check enforced that the domain being queried was part of active link with the domain specified in REMOTE_REQUESTER_HEADER. This is to ensure that the requesting user/domain has proper access to this data. In practice, this request is made by HQ directly, and this kind of check isn't necessary in that context so I'm assuming this check is strictly a security measure.

> I've replaced this check with the standard check for access to linked domains, which opens up the possibility that a user with linked domains permissions in both domains could retrieve data models from a domain that is not part of an active link with the requesting domain. I'm inclined to think this isn't a risk since the user has access to this information given their linked domain permission. Does this sound particularly concerning to anyone?

biyeun reviewed (approved):
hey @gherceg the changes made after offlining with Matt make sense, and the changes overall look good now. 👍

orangejenny commented:
> a user with linked domains permissions in both domains could retrieve data models from a domain that is not part of an active link with the requesting domain

I think this is fine, as you say, since the user has permissions in both domains.

I like the tests.
